packages/remix-server-runtime/serialize.ts

```diff
@@ -1,21 +1,60 @@
-import type { Jsonify } from "./jsonify";
+import type { EmptyObject, Jsonify } from "./jsonify";
 import type { TypedDeferredData, TypedResponse } from "./responses";
+import type {
+  ClientActionFunctionArgs,
+  ClientLoaderFunctionArgs,
+} from "./routeModules";
 import { expectType } from "./typecheck";
 import { type Expect, type Equal } from "./typecheck";
 
 // prettier-ignore
 /**
- * Infer JSON serialized data type returned by a loader or action.
+ * Infer JSON serialized data type returned by a loader or action, while
+ * avoiding deserialization if the input type if it's a clientLoader or
+ * clientAction that returns a non-Response
  *
  * For example:
  * `type LoaderData = SerializeFrom<typeof loader>`
  */
 export type SerializeFrom<T> =
-  T extends (...args: any[]) => infer Output ? Serialize<Awaited<Output>> :
+  T extends (...args: any[]) => infer Output ?
+    Parameters<T> extends [ClientLoaderFunctionArgs | ClientActionFunctionArgs] ?
+      // Client data functions may not serialize
+      SerializeClient<Awaited<Output>>
+    :
```

Comment:
Cleaned it up so that this is all that's new and we don't touch the flow for existing loaders/actions